### DISCUSSION TITLE
No more MicroSd support in 4.1

### DISCUSSION THREAD

**Discussion by kevinmcmurtrie** [osmandapp/OsmAnd#13539] [Q&A]
### Description
Google has completed their termination of microSd support in Android 11+.  Maps take 30 to 60 seconds EACH to index and panning across multiple maps incapacitates the system. A cold launch may take up to 15 minutes on a Snapdragon 888.

### How to reproduce
Stock Android 11 phone:
- Upgrade to OsmAnd+ beta in Play Store.
- Old workaround of changing the storage location no longer works due to permissions error.
- Move files back to original scoped storage location of microSd card
- OsmAnd+ runs about 10000 times slower; completely unusable.

### Your Environment
OsmAnd Version: 4.9 beta (guessing - had to downgrade)
Android/iOS version: 11
Device model: Xperia 1 III

**Maps used (online or offline):**   
All maps. Areas broken into many map regions are the worst.

**Comment by kevinmcmurtrie** [comment]
https://github.com/osmandapp/OsmAnd/issues/12046
https://github.com/osmandapp/OsmAnd/issues/10453

**Comment by sonora** [comment]
See also https://github.com/osmandapp/OsmAnd/issues/13219#issuecomment-976403727. We would have more leeway in non-PlayStore builds, e.g. could use targetSDK=29 and circumvent the issue at least for now. But it seems, understandably, non-PlayStore versions cannot be the focus.

Pretty amazing how google manages to cripple basic functionality and get away with it. :(

**Comment by vshcherb** [comment]
Just a note: this is slowness of sdcards doesn't happen on all phones with Android 11.

**Comment by Pastim** [comment]
In July, on Android 11, I was able to create a new osmand folder on my SD card and get osmand+ to use that with no major performance issues.  Have google changed their policy since then?  Is that why osmand+ 4.1 insists I move all my data?

**Comment by Pastim** [comment]
Help!

I have tried to create a new folder on main storage, and copied all sub-folders and most data files over from my SD card, but osmand+ says it can't find any osmand data.  I didn't copy all map data files because there isn't enough space.  

However, osmand says it can't find the data, so I am stuck.

**Comment by sonora** [comment]
Under many circumstances (i.e. for many storage locations), apps can now only "see" the files they have copied or created themselves. 

I have e.g. experimented with OsmAnd using a sub-folder of the SD-Card's "_Download_" like "_/storage/XXXX-XXXX/Downloads/osmand-data_", and that works fine, except OsmAnd itself needs to copy or create its files there. If you add anything like personal tracks or map obf's manually using e.g. a file manager app, these files will remain invisible to OsmAnd (It has to do with the Android Download.media library mechanism, I think).

This limitation should, however, not apply for folder _/Android/data/net.osmand(.plus)_ and it subfolders (I think), both on the "Internal" (non-.removable, in Osmand probably called "External 1") or "External (2)"/removable SD-Cards.

 @Pastim You should look at what you did in light of the above statements. I would re-start the OsmAnd app and then point it to an appropriate storage, i.e. let OsmAnd create its default folder in "External Storage 1". I think (not 100% sure) that once OsmAnd has created its correct subfolder of _/Android/data_ there, you can then also add the files there later manually via a file manager or a PC (via USB). (If this fails, maybe create a backup of your original map folder, then delete the maps which will not fit onto your new storage, and repeat the process using option "Let OsmAnd copy the files".)

To complicate matters: On all devices I have tested, /Android/data on External/removable SD-cards is not accessible from any Android file managers or PC at all. ;(  /Android/data on External/non-removable SD-memory is accessible from some Android file managers, and via USB.

_What a mess..._

**Comment by Pastim** [comment]
Indeed.  

I have, however, made some progress.  

I had to start afresh, which took quite a long time.  It's all a bit obscure.  On my Samsung A41 it turns out that it is possible to put the folder on the SD card in 'External Storage 2' (/storage/2614-1710/Android/data/net.osmand.plus), so after a lot of experimentation I selected this as my data storage folder.  The data within this can't be accessed from file managers on the mobile, nor by USB connected to my PC, which is a nuisance.  I can see the files in the folder via USB on my PC, but cannot copy them.  However, being on SD it does mean I have more space to download maps, contours, wikis and the rest.

It is then also possible to import maps from a shared folder, such as Downloads, using a file manager to 'open with' to osmand+.  Favourites can also be imported via 'My Places', so I haven't lost them all, which is a relief. Routes can be imported from Downloads via 'Configure map, 'Tracks', 'Add more'...' 

The next problem is exporting data.  I have, as yet, found no way to do that.   The main thing I regularly export is routes I have taken and recorded.  Is there any way to get at the 'rec' folder from outside OsmAnd+, or a command in OsmAnd+ to do so?  If not, I would ask for this to be considered as a feature in the next version of OsmAnd+.  The main reason I export such routes is that I can analyze them with gpx tools, see my average speed overall or part of a route, and so on.

Why these restrictions are imposed is a mystery to me.  Privacy?  That can't be right.  I can take photos, store them on shared and accessible SD, and export them, so why not maps, routes and so on?

**Comment by sonora** [comment]
Google in deed promotes data privacy improvement as the reason for this: Apps should essentially only see their own files, and nobody else should. There are a few specific exceptions, like well-defined import/export mechanisms via a file picker (which are not very user friendly or performant), or certain file types opened up for interoperability reasons via a library mechanism, like "Media", "Documents", or "Downloads".

On a practical note:
- Samsung's _MyFiles_ seems to date not be able to access Android/data/subfolders on either Ext1 nor Ext2 storage.
- AlphaInventor's _File Manager+_ can at least access Ext 1
- Lonely Cat Games' _Xplore File Manager_ is the only file manager app I have found so far which accesses Android/data/subfolders also on Ext 2.
- I believe this is done by temporarily renaming "Android" to something else (e.g. Android1). If you need emergency access without third party tool, e.g. via USB, you can do just that manually (with the adverse effects which may result from removing the Android folder from the visibility of all -potentially running- apps on the system...).

**Comment by Pastim** [comment]
Agreed.  

Samsung MyFiles, fx Explorer, and Google Files can't see the content of any of the data subfolders on Ext 2.  However, I have another Files app (Marc apps & software) which can see the files, and even tries to copy them, but fails.  It also has a 'share' option.  I tried that sharing to email, and the weird thing is it did share most of a route file, but dropped the last 10 or so lines (i tried on several files).  

I think I won't try renaming things just yet, I wonder if it's possible to create a link from somewhere accessible.  I know how to do that on linux, so it might be possible, but I know little of Android.

**Comment by brunoais** [comment]
@sonora In reality, Google could have implemented the same kind of security with non-inherited mounts and permissions.
However, they decided not to. They decided to use a much more (unecessarily) complicated method.
Linux 5.X already comes with very advanced mounting options which would solve all this debacle. If you want to know how far it can go, you can see what docker or even podman allows by leveraging Linux features.
Do note it IS possible to make Android safe by having a rooted app that gates all other apps by providing mounts paths as requested by the apps with the acceptance of the user as appropriate given their the permissions they are supposed to have.
It would be much simpler to implement and multiple times more efficient

**Comment by kevinmcmurtrie** [comment]
Has anyone found a fix for OsmAnd+ 4.19 under Android 11?  I upgraded my older phone for experimentation but I can't find any way to make it work.  No matter what permission I grant OsmAnd+ in ADB, anything outside of scoped storage on the microSD card gives an access error.  I tried old permissions and I tried new Android 11 permissions.  When I put the maps in scoped storage, OsmAnd+ only shows a blank map because SAF is insanely so slow.  "Download Maps" : "LOCAL" eventually shows my maps but nothing else ever finishes loading even after 25 minutes.

Must I keep my main phone on OsmAnd+ 4.08 forever?

Phones:  T-Mobile LG Velvet 5G and Sony Xperia 1 III

**Comment by Pastim** [comment]
My files are now in /storage/2614-1710/Android/data/net.osmand.plus, on my SD card.  OsmAnd+ calls this 'External storage 2'.  Whether this is 'scoped storage' or not I don't know - I've failed to understand what the term means.

OsmAnd+ now runs OK, maybe a little slower, but perfectly usable.  The mobile is a Samsung A41.  The hardest thing for me was to find ways to import and export maps, routes, favourites and so on, but I've now found ways to do all those things.  It's just a little more inconvenient than it used to be when I could just connect my PC via USB and read and write files using the PC file manager.

**Comment by kevinmcmurtrie** [comment]
> My files are now in /storage/2614-1710/Android/data/net.osmand.plus, on my SD card. OsmAnd+ calls this 'External storage 2'. Whether this is 'scoped storage' or not I don't know - I've failed to understand what the term means.
> 
> OsmAnd+ now runs OK, maybe a little slower, but perfectly usable. The mobile is a Samsung A41. The hardest thing for me was to find ways to import and export maps, routes, favourites and so on, but I've now found ways to do all those things. It's just a little more inconvenient than it used to be when I could just connect my PC via USB and read and write files using the PC file manager.

I heard that some phones have a built-in workaround for Google's SAF.  There's no such luck with mine.

Your storage type is what I must use in 4.19.  My LG phone finally showed a map after running for several hours.  It then worked with light use.  Panning or pinch zooming across map regions kills rendering for at least a minute.  Searching is really slow.  The indexing after downloading a new map has gone from a few seconds to ... and hour maybe?    It's the difference between running smoothly and being unusable.

**Comment by tommycrock** [comment]
#12982

**Comment by kevinmcmurtrie** [comment]
Tried on two Android 11 phones:
`sm set-isolated-storage off` - Not supported anymore.  Google's cursed FUSE app crashes, the microSd card fails to mount, and just looking at the volume list hangs for several seconds.  You can't use any microSd card until this setting is reverted.

`sm partition <DISK> private` - OsmAnd can select this storage and it seems fast!  **BUT**, adoptable storage is not supported. The union mount is missing and the card is completely invisible except for the one unique scoped storage option seen by some apps.  Even ADB sees nothing more of it than a mount path in `df`.

**Comment by Pastim** [comment]
It was running OK (using /storage/2614-1710/Android/data/net.osmand.plus), but I then recently found 'shared' route gpxs were being truncated when sent by email or attempting to copy to shareable storage (this was working up and until only a few days ago).   So I tried  /storage/2614-1710/Download/Osmand, and in 4.1.11 it now works as it used to, all files accessible externally.  Good!

**Comment by kevinmcmurtrie** [comment]
> So I tried /storage/2614-1710/Download/Osmand, and in 4.1.11 it now works as it used to, all files accessible externally. Good!

Can you still download maps or pan across regions?  I tried using Download and Documents on my test phone.  OsmAnd+ crashes and gives nonsense I/O errors constantly.  That's on top of all the usual Google SAF stalls and hangs of scoped storage.
🖕 Google

**Comment by tommycrock** [comment]
> > So I tried /storage/2614-1710/Download/Osmand, and in 4.1.11 it now works as it used to, all files accessible externally. Good!
> 
> Can you still download maps or pan across regions? I tried using Download and Documents on my test phone. OsmAnd+ crashes and gives nonsense I/O errors constantly. That's on top of all the usual Google SAF stalls and hangs of scoped storage. 🖕 Google

I can still download maps and pan across regions using download (although in '0', device memory). I've noticed that things don't work immediately after starting the phone - either it triggers the "get started" wizard or Osmand loads and I can see favourites and hill-shading but maps don't work. Leave the phone a minute or so and it all seems to work as well as it did before for me. I've not tried on an SD card

**Comment by Pastim** [comment]
> > So I tried /storage/2614-1710/Download/Osmand, and in 4.1.11 it now works as it used to, all files accessible externally. Good!
> 
> Can you still download maps or pan across regions? I tried using Download and Documents on my test phone. OsmAnd+ crashes and gives nonsense I/O errors constantly. That's on top of all the usual Google SAF stalls and hangs of scoped storage. middle_finger Google

All fine for me.  I've downloaded another map and associated data, recorded a route, panned across UK regions, all running fine on my SD card, Samsung A41, Android 11.

**Comment by kevinmcmurtrie** [comment]
Today Google updated my OsmAnd+ without my permission - it's permanently destroyed.  OsmAnd from osmand.net is the limited version.  OsmAnd~ from F-Droid _seems_ to have all the paid features but it can't sync to the original Play Store purchase or the OsmLive subscription.

Is there an official way to have the paid features without Play Store problems?  Amazon store?  Is F-Droid good enough?

**Comment by pebogufi** [comment]
For this situation it is too late, but my recommandation is:
do NEVER allow Google to make any AUTOMATIC updates. 
You can switch it off in settings in Play Store.
Here 
https://download.osmand.net/releases/
you can dnl older versions and in most cases it works to install a older one over a newer one.

**Comment by Pastim** [comment]
> Today Google updated my OsmAnd+ without my permission - it's permanently destroyed. OsmAnd from osmand.net is the limited version. OsmAnd~ from F-Droid _seems_ to have all the paid features but it can't sync to the original Play Store purchase or the OsmLive subscription.
> 
> Is there an official way to have the paid features without Play Store problems? Amazon store? Is F-Droid good enough?
 

What version of Android are you on?  What version OsmAnd+ was downloaded?  I'm using the Google Play version and it's working for me. 

When you say 'permanently destroyed', I assume you mean it won't run at all, whatever you do.  Is that correct?  A while ago I had to faff about moving the data until it would work, but it did work in the end, so 'permanently destroyed' sounds pretty extreme.

**Comment by kevinmcmurtrie** [comment]
Yesterday Google replaced my OsmAnd 4.0.9 with 4.1.11 without permission and while automatic updates are turned off.

I have two different makes of Android 11 phones (Sony, LG) and neither can use the microSD card with Play Store apps released after Nov 2021.  Apps are slow, hang, crash, and get I/O errors using it.  Google's own Files app fails on microSD cards after it was updated.  There doesn't even appear to be read-after-write consistency!

I've replaced some Play Store apps with F-Droid apps that avoid Google's broken SAF.  OsmAnd+ is a bit more complicated because it's integrated with Play Store purchases.  OsmAnd~ 4.0.9 on F-Droid seems to work so far but it can't sync on purchases and I can't upgrade there either.

**Comment by Pastim** [comment]
> Yesterday Google replaced my OsmAnd 4.0.9 with 4.1.11 without permission and while automatic updates are turned off.
> 
> I have two different makes of Android 11 phones (Sony, LG) and neither can use the microSD card with Play Store apps released after Nov 2021. Apps are slow, hang, crash, and get I/O errors using it. Google's own Files app fails on microSD cards after it was updated. There doesn't even appear to be read-after-write consistency!
> 
> I've replaced some Play Store apps with F-Droid apps that avoid Google's broken SAF. OsmAnd+ is a bit more complicated because it's integrated with Play Store purchases. OsmAnd~ 4.0.9 on F-Droid seems to work so far but it can't sync on purchases and I can't upgrade there either.

You have some very strange things going on.  I never, ever, get such updates without explicit permission. Now I am using the SD Card Download directory for OsmAnd+ I have no problems.  When using the Android/data/ directory few file manager apps can access the files, and that seems to be deliberate, rather than failure.

**Comment by kevinmcmurtrie** [comment]
I tried the Download and Documents directory (see above).  It's just as slow as scoped storage but adds crashes and I/O errors due to not having read-after-write consistency.  This isn't specific to OsmAnd or the phone, its that Google's SAF is THAT bad.  https://issuetracker.google.com/ is full of bug reports but Googlers think they can do no wrong so they keep closing them.

My phone is a Sony Xperia 1 III so I don't have the bug workarounds that Samsung has added to theirs.  My older MediaTek LG Velvet 5G shows exactly the same bugs so I've been using it for experimentation.

Yeah, I know this is a Google issue in an OsmAnd ticketing system.  I'm hoping that a workaround can be found.  I'd be willing to buy OsmAnd+ in a different app store if that meant I could resume software updates without microSD issues.

**Comment by Pastim** [comment]
There is still a problem.   I can use USB or a file manager to put files in the tracks folder, but OsmAnd+ cannot see them. However. I can add gpx files to the SD Card Download directory and import tracks from there. I am not sure I'll be able to copy maps to OsmAnd+.

**Comment by tommycrock** [comment]
> There is still a problem. I can use USB or a file manager to put files in the tracks folder, but OsmAnd+ cannot see them. However. I can add gpx files to the SD Card Download directory and import tracks from there. I am not sure I'll be able to copy maps to OsmAnd+.

Yes. The download folder workaround essentially only gives you read access (you can write files but they won't be read and get in the way - there may be some exceptions). 

You can import maps via a file manager using 'open with' -> osmand, one at a time.

**Comment by sonora** [comment]
The Download folders seems to follow a library approach: While many apps and mechanisms can write to it, only few apps (mostly file managers) 'see' **all** files in it. Other than using the Open with or file picker dialogs, by my experience OsmAnd can now only see exactly the subset of files it placed there itself, see https://github.com/osmandapp/OsmAnd/issues/12046#issuecomment-1005510733.

**Comment by sonora** [comment]
@vshcherb I wonder if under the circumstances we should simply use the Android "All files access" permission reserved for exceptional cases (https://developer.android.com/training/data-storage/manage-all-files#all-files-access-google-play) and see if the Play store has any objections at all?

Reasoning is that OsmAnd is essentially a "Document management app" for maps, tracks, locations,  travel-wikis, trip recordings, etc.

**Reply by vshcherb** [reply]
Moved to discussion https://github.com/osmandapp/OsmAnd/discussions/13538

**Comment by vshcherb** [comment]
I wouldn't do it we are definitely not in this business and neither of our competitors has this permission. Of course we could distribute this version via other markets but for sure not via GPlay. I've been having blocked OsmAnd for 4 times and that's not a pleasant experience even if user could continue download it.

**Comment by j-schneid** [comment]
Placing Osmand in the download folder is a bit of an awkward workaround but better than nothing. Now it would be very helpful if there was a way to import _multiple_ files (e.g. tracks) at a time. I have got a collection of tracks of several years that I would like to be available in Osmand. Unfortunately I have not exported them with Osmand's own method into one single file when this was still possible. Now I have a migrated Osmand that has (for whatever reason) left it's tracks behind and is kind of "empty" for me. I have got them all on my computer, but importing them one by one is not an option. (Same probably goes for av notes, they are just not so important for me personally.)

**Reply by mariush444** [reply]
you can try to use SD_card/Android/media/net.osmand.plus
someone mentioned this on Telegram channel and it works for me too.

**Reply by j-schneid** [reply]
Wow, great! Thank you so much, this is a solution. Here I can copy back all the files from my computer and osmand can see them. And Foldersync can access it, too, so my regular backup is also working again.
